File: morris/ui.py
from __future__ import annotations

import math
import time
from dataclasses import dataclass
from typing import Optional, Tuple, List

import pygame

from .constants import POINTS_NORM, PLAYER_WHITE, PLAYER_BLACK, EMPTY, ADJACENT
from .game_state import GameState, Move
from .ai import choose_move, EASY, MEDIUM, HARD, AIDifficulty, hint as ai_hint
from .stats import load_stats, save_stats, achievement_text, Stats
from .puzzles import load_puzzle, SAMPLE_PUZZLES


@dataclass
class UIConfig:
	width: int = 800
	height: int = 800
	margin: int = 60
	bg_color: Tuple[int, int, int] = (18, 18, 24)
	board_color: Tuple[int, int, int] = (200, 200, 200)
	white_color: Tuple[int, int, int] = (240, 240, 240)
	black_color: Tuple[int, int, int] = (20, 20, 20)
	highlight_color: Tuple[int, int, int] = (255, 215, 0)
	ghost_color: Tuple[int, int, int] = (100, 100, 120)
	text_color: Tuple[int, int, int] = (230, 230, 240)
	point_radius: int = 16
	selected_radius: int = 22
	anim_speed_px_per_s: float = 800.0


class UI:
	def __init__(self, state: GameState, ai_enabled: bool = False, ai_color: int = PLAYER_BLACK, difficulty: AIDifficulty = MEDIUM, config: Optional[UIConfig] = None) -> None:
		self.state = state
		self.ai_enabled = ai_enabled
		self.ai_color = ai_color
		self.difficulty = difficulty
		self.config = config or UIConfig()
		self._selected: Optional[int] = None
		self._anim: Optional[Tuple[int, int, float, float]] = None  # from_idx, to_idx, start_time, duration
		self._hint_move: Optional[Move] = None
		self._stats = load_stats()
		self._last_mill_count = 0
		self._dragging: bool = False
		self._drag_from: Optional[int] = None
		self._legal_targets: List[int] = []

		pygame.init()
		self.screen = pygame.display.set_mode((self.config.width, self.config.height))
		pygame.display.set_caption("Nine Men's Morris")
		self.clock = pygame.time.Clock()
		self.font = pygame.font.SysFont("Arial", 20)

	def norm_to_px(self, x: float, y: float) -> Tuple[int, int]:
		mx, my = self.config.margin, self.config.margin
		w = self.config.width - 2 * mx
		h = self.config.height - 2 * my
		return int(mx + x * w), int(my + y * h)

	def point_pos(self, idx: int) -> Tuple[int, int]:
		nx, ny = POINTS_NORM[idx]
		return self.norm_to_px(nx, ny)

	def nearest_point(self, pos: Tuple[int, int]) -> Optional[int]:
		x, y = pos
		for i in range(24):
			px, py = self.point_pos(i)
			if math.hypot(px - x, py - y) <= self.config.point_radius * 1.6:
				return i
		return None

	def draw_board(self) -> None:
		self.screen.fill(self.config.bg_color)
		c = self.config.board_color
		# Draw lines between adjacent points
		for i in range(24):
			for j in ADJACENT[i]:
				if j > i:
					pygame.draw.line(self.screen, c, self.point_pos(i), self.point_pos(j), 2)
		# Draw points and pieces
		for i in range(24):
			px, py = self.point_pos(i)
			pygame.draw.circle(self.screen, c, (px, py), self.config.point_radius, 2)
			v = self.state.board[i]
			if v != EMPTY:
				color = self.config.white_color if v == PLAYER_WHITE else self.config.black_color
				pygame.draw.circle(self.screen, color, (px, py), self.config.point_radius - 3)
			# highlight legal target
			if i in self._legal_targets:
				pygame.draw.circle(self.screen, self.config.highlight_color, (px, py), self.config.point_radius + 2, 2)
		# Draw selection highlight
		if self._selected is not None:
			px, py = self.point_pos(self._selected)
			pygame.draw.circle(self.screen, self.config.highlight_color, (px, py), self.config.selected_radius, 2)
		# Draw hint arrow
		if self._hint_move and self._hint_move.to_idx is not None:
			fx = self._hint_move.from_idx if self._hint_move.from_idx is not None else self._hint_move.to_idx
			if fx is not None:
				p1 = self.point_pos(fx)
				p2 = self.point_pos(self._hint_move.to_idx)
				pygame.draw.line(self.screen, self.config.highlight_color, p1, p2, 3)

		# Status text
		turn = "White" if self.state.to_move == PLAYER_WHITE else "Black"
		phase = self.state.phase + (" (remove)" if self.state.removal_pending else "")
		msg = f"Turn: {turn} | Phase: {phase} | W:{self.state.white_in_hand} B:{self.state.black_in_hand}"
		txt = self.font.render(msg, True, self.config.text_color)
		self.screen.blit(txt, (10, 10))
		ach_txt = achievement_text(self._stats)
		if ach_txt:
			badge = self.font.render(ach_txt, True, self.config.highlight_color)
			self.screen.blit(badge, (10, 36))

	def animate_move(self, from_idx: int, to_idx: int) -> None:
		p1 = self.point_pos(from_idx)
		p2 = self.point_pos(to_idx)
		distance = math.hypot(p2[0] - p1[0], p2[1] - p1[1])
		dur = max(0.08, distance / self.config.anim_speed_px_per_s)
		self._anim = (from_idx, to_idx, time.time(), dur)

	def update_animation(self) -> None:
		if not self._anim:
			return
		from_idx, to_idx, start, dur = self._anim
		alpha = (time.time() - start) / dur
		if alpha >= 1.0:
			self._anim = None
			return
		# Draw moving ghost
		x1, y1 = self.point_pos(from_idx)
		x2, y2 = self.point_pos(to_idx)
		x = int(x1 + (x2 - x1) * alpha)
		y = int(y1 + (y2 - y1) * alpha)
		color = self.config.white_color if self.state.to_move == PLAYER_BLACK else self.config.black_color
		pygame.draw.circle(self.screen, self.config.ghost_color, (x, y), self.config.point_radius - 3)

	def handle_click(self, pos: Tuple[int, int]) -> None:
		if self.state.winner is not None:
			return
		idx = self.nearest_point(pos)
		if idx is None:
			self._selected = None
			return
		# Removal phase: choose opponent piece to remove
		if self.state.removal_pending:
			move = Move(player=self.state.to_move, remove_idx=idx)
			self._apply_and_record(move)
			self._selected = None
			return
		# Placing phase
		if self.state.phase == "placing":
			move = Move(player=self.state.to_move, to_idx=idx)
			if self._apply_and_record(move):
				self._selected = None
			return
		# Moving/flying
		if self._selected is None:
			if self.state.board[idx] == self.state.to_move:
				self._selected = idx
				self._compute_legal_targets(idx)
			return
		else:
			from_idx = self._selected
			move = Move(player=self.state.to_move, from_idx=from_idx, to_idx=idx)
			if self._apply_and_record(move):
				# Animate from -> to
				self.animate_move(from_idx, idx)
				self._selected = None
				self._legal_targets = []
				return
			# If failed, maybe reselect
			if self.state.board[idx] == self.state.to_move:
				self._selected = idx
				self._compute_legal_targets(idx)
			else:
				self._selected = None
				self._legal_targets = []

	def _compute_legal_targets(self, from_idx: int) -> None:
		self._legal_targets = []
		for mv in self.state.legal_moves():
			if mv.from_idx == from_idx:
				if mv.to_idx is not None:
					self._legal_targets.append(mv.to_idx)

	def _apply_and_record(self, move: Move) -> bool:
		before_to_move = self.state.to_move
		ok = self.state.apply_move(move)
		if not ok:
			return False
		# Stats via events
		for ev in self.state.events:
			if ev["type"] in ("placed", "moved"):
				self._stats.record_move()
			elif ev["type"] == "formed_mill":
				self._stats.record_mill()
				save_stats(self._stats)
			elif ev["type"] == "game_over":
				self._stats.record_win(ev["winner"])
				save_stats(self._stats)
		return True

	def maybe_ai_move(self) -> None:
		if not self.ai_enabled:
			return
		if self.state.winner is not None:
			return
		if self.state.to_move != self.ai_color:
			return
		mv = choose_move(self.state, self.difficulty)
		if mv:
			from_idx = mv.from_idx
			to_idx = mv.to_idx
			self._apply_and_record(mv)
			if from_idx is not None and to_idx is not None:
				self.animate_move(from_idx, to_idx)

	def set_difficulty(self, key: int) -> None:
		self.difficulty = {pygame.K_e: EASY, pygame.K_m: MEDIUM, pygame.K_h: HARD}.get(key, self.difficulty)

	def loop(self) -> None:
		running = True
		while running:
			for event in pygame.event.get():
				if event.type == pygame.QUIT:
					running = False
				elif event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
					# Start drag if in moving/flying and clicked own piece; otherwise handle as click
					idx = self.nearest_point(event.pos)
					if idx is not None and self.state.phase != "placing" and not self.state.removal_pending and self.state.board[idx] == self.state.to_move:
						self._dragging = True
						self._drag_from = idx
						self._selected = idx
						self._compute_legal_targets(idx)
					else:
						self.handle_click(event.pos)
				elif event.type == pygame.MOUSEBUTTONUP and event.button == 1 and self._dragging:
					idx = self.nearest_point(event.pos)
					if idx is not None and self._drag_from is not None:
						mv = Move(player=self.state.to_move, from_idx=self._drag_from, to_idx=idx)
						if self._apply_and_record(mv):
							self.animate_move(self._drag_from, idx)
					self._selected = None
					self._legal_targets = []
					self._dragging = False
					self._drag_from = None
				elif event.type == pygame.KEYDOWN:
					if event.key == pygame.K_u:
						self.state.undo()
					elif event.key == pygame.K_r:
						self.state.redo()
					elif event.key in (pygame.K_e, pygame.K_m, pygame.K_h):
						self.set_difficulty(event.key)
					elif event.key == pygame.K_h:
						self._hint_move = ai_hint(self.state, self.difficulty)
					elif event.key == pygame.K_n:
						# new game
						self.state.reset()
						self._selected = None
						self._hint_move = None
						self._legal_targets = []
						self._dragging = False
						self._drag_from = None
					elif event.key == pygame.K_a:
						# toggle AI
						self.ai_enabled = not self.ai_enabled
					elif event.key == pygame.K_p:
						# load random puzzle
						from random import randint
						idx = randint(0, len(SAMPLE_PUZZLES) - 1)
						p = load_puzzle(idx)
						if p:
							self.state = p
							self._selected = None
							self._hint_move = None
							self._legal_targets = []
							self._dragging = False
							self._drag_from = None
			self.maybe_ai_move()
			self.draw_board()
			self.update_animation()
			if self.state.winner is not None:
				msg = "White wins!" if self.state.winner == PLAYER_WHITE else "Black wins!"
				banner = self.font.render(msg, True, self.config.highlight_color)
				rect = banner.get_rect(center=(self.config.width // 2, 30))
				self.screen.blit(banner, rect)
			pygame.display.flip()
			self.clock.tick(60)
		pygame.quit()

File: web/js/ui.js
import { POINTS_NORM, PLAYER_WHITE, PLAYER_BLACK, EMPTY, ADJACENT } from './constants.js';
import { Move } from './state.js';
import { chooseMove, hint as aiHint, EASY, MEDIUM, HARD } from './ai.js';

function createSVG(tag, attrs={}){
  const el = document.createElementNS('http://www.w3.org/2000/svg', tag);
  for(const [k,v] of Object.entries(attrs)) el.setAttribute(k, String(v));
  return el;
}

export class GameUI{
  constructor(state, mount){
    this.state = state;
    this.mount = mount;
    this.width = 1000; this.height = 1000;
    this.aiEnabled = false; this.aiColor = PLAYER_BLACK; this.difficulty = MEDIUM;
    this.selected = null; this.legalTargets = [];
    this.listeners = new Set();
    this.showHint = false; this.hintMove = null;
    this.svg = null; this.piecesLayer = null;
    this._setup();
  }
  onChange(fn){ this.listeners.add(fn); return () => this.listeners.delete(fn); }
  emitChange(){ for(const fn of this.listeners) fn(); }
  setAIEnabled(v){ this.aiEnabled = v; this.emitChange(); }
  setAIColor(c){ this.aiColor = c; this.emitChange(); }
  setAIDifficulty(d){ this.difficulty = d; this.emitChange(); }
  toggleHint(){ this.showHint = !this.showHint; if(!this.showHint) this.hintMove=null; this.render(); }
  clearSelections(){ this.selected=null; this.legalTargets=[]; this.hintMove=null; }

  _setup(){
    this.mount.innerHTML = '';
    const svg = createSVG('svg', { class: 'board-svg', viewBox: '0 0 1000 1000' });
    this.svg = svg;
    const edgesLayer = createSVG('g');
    const pointsLayer = createSVG('g');
    const piecesLayer = createSVG('g');
    this.piecesLayer = piecesLayer;

    // Edges from ADJACENT
    for(let i=0;i<24;i++){
      for(const j of ADJACENT[i]){
        if(j>i){
          const [x1,y1] = this._pointPx(i); const [x2,y2] = this._pointPx(j);
          edgesLayer.appendChild(createSVG('line', { class:'edge', x1,y1,x2,y2 }));
        }
      }
    }

    // Points
    for(let i=0;i<24;i++){
      const [cx,cy] = this._pointPx(i);
      const p = createSVG('circle', { class:'point', r:16, cx, cy, 'data-idx': i });
      p.addEventListener('click', () => this._handlePointClick(i));
      pointsLayer.appendChild(p);
    }

    svg.appendChild(edgesLayer);
    svg.appendChild(pointsLayer);
    svg.appendChild(piecesLayer);
    this.mount.appendChild(svg);

    // Drag support
    svg.addEventListener('mousedown', (e) => this._onMouseDown(e));
    svg.addEventListener('mousemove', (e) => this._onMouseMove(e));
    window.addEventListener('mouseup', (e) => this._onMouseUp(e));
  }

  _pointPx(idx){
    const [nx,ny] = POINTS_NORM[idx];
    return [60 + nx * (1000-120), 60 + ny * (1000-120)];
  }

  _handlePointClick(idx){
    if(this.state.winner!==null) return;
    if(this.state.removalPending){
      const mv = new Move(this.state.toMove, null, null, idx);
      if(this.state.applyMove(mv)){
        this.selected=null; this.legalTargets=[]; this._afterMove();
      }
      return;
    }
    if(this.state.phase==='placing'){
      const mv = new Move(this.state.toMove, null, idx, null);
      if(this.state.applyMove(mv)){
        this.selected=null; this.legalTargets=[]; this._afterMove();
      }
      return;
    }
    if(this.selected===null){
      if(this.state.board[idx]===this.state.toMove){
        this.selected = idx; this._computeLegalTargets(idx);
      }
    } else {
      const fromIdx = this.selected;
      const mv = new Move(this.state.toMove, fromIdx, idx, null);
      if(this.state.applyMove(mv)){
        this.selected=null; this.legalTargets=[]; this._afterMove();
      } else if(this.state.board[idx]===this.state.toMove){
        this.selected=idx; this._computeLegalTargets(idx);
      } else {
        this.selected=null; this.legalTargets=[];
      }
    }
    this.render();
  }

  _computeLegalTargets(fromIdx){
    this.legalTargets = [];
    for(const mv of this.state.legalMoves()){
      if(mv.fromIdx===fromIdx && mv.toIdx!==null) this.legalTargets.push(mv.toIdx);
    }
  }

  _onMouseDown(e){
    if(this.state.phase==='placing' || this.state.removalPending || this.state.winner!==null) return;
    const idx = this._hitPoint(e);
    if(idx!==null && this.state.board[idx]===this.state.toMove){
      this.selected = idx; this._computeLegalTargets(idx); this.render();
    }
  }
  _onMouseMove(e){
    // could draw ghost piece while dragging in future
  }
  _onMouseUp(e){
    if(this.selected===null) return;
    const idx = this._hitPoint(e);
    if(idx!==null){
      const mv = new Move(this.state.toMove, this.selected, idx, null);
      if(this.state.applyMove(mv)){ this.selected=null; this.legalTargets=[]; this._afterMove(); this.render(); return; }
    }
    this.selected=null; this.legalTargets=[]; this.render();
  }
  _hitPoint(e){
    const pt = this.svg.createSVGPoint(); pt.x = e.offsetX; pt.y = e.offsetY;
    // hit test by distance to each point
    for(let i=0;i<24;i++){
      const [cx,cy] = this._pointPx(i);
      const dx = pt.x-cx, dy=pt.y-cy; if(Math.hypot(dx,dy) <= 22) return i;
    }
    return null;
  }

  _afterMove(){
    for(const ev of this.state.events){
      if(ev.type==='formed_mill'){ this._toast('Mill formed! Remove an opponent piece.'); this._pulseTargets(); }
      if(ev.type==='game_over'){
        this._celebrate(ev.winner);
      }
    }
    this.emitChange();
    this.maybeAIMove();
  }

  maybeAIMove(){
    if(!this.aiEnabled) return;
    if(this.state.winner!==null) return;
    if(this.state.toMove!==this.aiColor) return;
    const mv = chooseMove(this.state, this.difficulty);
    if(mv){ this.state.applyMove(mv); this.emitChange(); this.render(); }
  }

  setToastContainer(){
    if(!this.toast){ this.toast = document.createElement('div'); this.toast.className='toast'; this.mount.appendChild(this.toast); }
  }
  _toast(msg){ this.setToastContainer(); this.toast.textContent = msg; clearTimeout(this.toastTimer); this.toastTimer = setTimeout(()=>{ this.toast.remove(); this.toast=null; }, 2000); }

  _pulseTargets(){
    // add a brief class to highlight target points
    const points = this.svg.querySelectorAll('circle.point');
    points.forEach(p => { p.classList.remove('target'); });
  }

  _celebrate(winner){
    const container = document.getElementById('confettiContainer');
    for(let i=0;i<120;i++){
      const el = document.createElement('div');
      el.className='confetti';
      const hue = winner===PLAYER_WHITE ? 210 : 0;
      el.style.background = `hsl(${hue + Math.random()*30}, 90%, ${50 + Math.random()*20}%)`;
      el.style.left = Math.random()*100 + 'vw';
      el.style.animation = `confetti-fall ${2+Math.random()*2}s linear ${Math.random()*0.6}s forwards`;
      container.appendChild(el);
      setTimeout(()=> el.remove(), 4000);
    }
    this._toast(winner===PLAYER_WHITE ? 'White wins!' : 'Black wins!');
  }

  render(){
    // Points styling
    const pointEls = this.svg.querySelectorAll('circle.point');
    pointEls.forEach((p) => {
      const idx = Number(p.getAttribute('data-idx'));
      p.classList.toggle('selected', this.selected===idx);
      p.classList.toggle('target', this.legalTargets.includes(idx) || (this.state.phase==='placing' && this.state.board[idx]===EMPTY));
    });

    // Pieces
    this.piecesLayer.innerHTML = '';
    for(let i=0;i<24;i++){
      const v = this.state.board[i]; if(v===EMPTY) continue;
      const [cx,cy] = this._pointPx(i);
      const piece = createSVG('circle', { class: 'piece ' + (v===PLAYER_WHITE?'white':'black'), r:13, cx, cy });
      this.piecesLayer.appendChild(piece);
    }

    // Hint arrow
    if(this.showHint){
      this.hintMove = aiHint(this.state, this.difficulty);
      if(this.hintMove && this.hintMove.toIdx!==null){
        const from = this.hintMove.fromIdx ?? this.hintMove.toIdx;
        const to = this.hintMove.toIdx;
        const [x1,y1] = this._pointPx(from); const [x2,y2] = this._pointPx(to);
        const arrow = createSVG('line', { x1, y1, x2, y2, stroke: '#ffd700', 'stroke-width': 6, 'stroke-linecap':'round' });
        this.piecesLayer.appendChild(arrow);
      }
    }
  }
}
